File: apps/android/app/src/main/java/com/ailimits/dashboard/provider/AnthropicProvider.kt
package com.ailimits.dashboard.provider

import com.ailimits.dashboard.data.ProviderConfig
import com.ailimits.dashboard.data.UsageSnapshot
import com.ailimits.dashboard.data.UsageStatus
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.URL
import javax.net.ssl.HttpsURLConnection

class AnthropicProvider : UsageProvider {
    override fun validateConfig(config: ProviderConfig): Boolean =
        config.type == "anthropic" && !config.apiKey.isNullOrBlank()

    override suspend fun fetchUsage(config: ProviderConfig): UsageSnapshot =
        withContext(Dispatchers.IO) {
            if (!validateConfig(config) || config.apiKey == null) return@withContext errorSnapshot(config.id)
            val now = java.util.Calendar.getInstance()
            val startOfMonth = java.util.Calendar.getInstance().apply {
                set(java.util.Calendar.DAY_OF_MONTH, 1)
                set(java.util.Calendar.HOUR_OF_DAY, 0)
                set(java.util.Calendar.MINUTE, 0)
                set(java.util.Calendar.SECOND, 0)
                set(java.util.Calendar.MILLISECOND, 0)
            }.time
            val endOfMonth = now.time
            val urlStr =
                "https://api.anthropic.com/v1/organizations/cost_report?starting_at=${java.net.URLEncoder.encode(java.text.SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", java.util.Locale.US).apply { timeZone = java.util.TimeZone.getTimeZone("UTC") }.format(startOfMonth), "UTF-8")}&ending_at=${java.net.URLEncoder.encode(java.text.SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", java.util.Locale.US).apply { timeZone = java.util.TimeZone.getTimeZone("UTC") }.format(endOfMonth), "UTF-8")}&bucket_width=1d&limit=31"
            try {
                val url = URL(urlStr)
                (url.openConnection() as HttpsURLConnection).apply {
                    requestMethod = "GET"
                    setRequestProperty("x-api-key", config.apiKey!!)
                    setRequestProperty("anthropic-version", "2023-06-01")
                    setRequestProperty("Content-Type", "application/json")
                    connectTimeout = 10000
                    readTimeout = 10000
                }.run {
                    if (responseCode != 200) return@withContext errorSnapshot(config.id)
                    val body = inputStream.bufferedReader().use { it.readText() }
                    val json = JSONObject(body)
                    val data = json.optJSONArray("data") ?: return@withContext errorSnapshot(config.id)
                    var used = 0.0
                    for (i in 0 until data.length()) {
                        val bucket = data.optJSONObject(i) ?: continue
                        val results = bucket.optJSONArray("results") ?: continue
                        for (j in 0 until results.length()) {
                            val r = results.optJSONObject(j)
                            used += r?.optDouble("amount", 0.0) ?: 0.0
                        }
                    }
                    val limit = config.manualLimit ?: 100.0
                    val ratio = if (limit > 0) used / limit else 0.0
                    UsageSnapshot(
                        providerId = config.id,
                        used = used,
                        limit = limit,
                        unit = "USD",
                        timestamp = System.currentTimeMillis(),
                        status = usageStatus(ratio),
                        displayName = config.displayName ?: "Anthropic",
                    )
                }
            } catch (_: Exception) {
                errorSnapshot(config.id)
            }
        }

    private fun errorSnapshot(providerId: String) = UsageSnapshot(
        providerId = providerId,
        used = 0.0,
        limit = 1.0,
        unit = "USD",
        timestamp = System.currentTimeMillis(),
        status = UsageStatus.ERROR,
        displayName = "Anthropic",
    )
}

File: packages/core/src/providers/AnthropicProvider.ts
import type { ProviderConfig } from "../models/ProviderConfig.js";
import type { UsageSnapshot } from "../models/UsageSnapshot.js";
import type { IUsageProvider } from "./IUsageProvider.js";

const API_BASE = "https://api.anthropic.com/v1";
const DEFAULT_LIMIT = 100; // $100 в месяц по умолчанию

interface CostResult {
  amount?: string;
  currency?: string;
}

interface CostBucket {
  results?: CostResult[];
}

interface CostReportResponse {
  data?: CostBucket[];
}

export class AnthropicProvider implements IUsageProvider {
  validateConfig(config: ProviderConfig): boolean {
    return config.type === "anthropic" && !!config.apiKey;
  }

  async fetchUsage(config: ProviderConfig): Promise<UsageSnapshot> {
    if (!this.validateConfig(config) || !config.apiKey) {
      return this.errorSnapshot(config.id);
    }
    const now = new Date();
    const startOfMonth = new Date(now.getFullYear(), now.getMonth(), 1);
    const startingAt = startOfMonth.toISOString();
    const endingAt = now.toISOString();
    const url = `${API_BASE}/organizations/cost_report?starting_at=${encodeURIComponent(startingAt)}&ending_at=${encodeURIComponent(endingAt)}&bucket_width=1d&limit=31`;
    try {
      const res = await fetch(url, {
        headers: {
          "x-api-key": config.apiKey,
          "anthropic-version": "2023-06-01",
          "Content-Type": "application/json",
        },
      });
      if (!res.ok) {
        await res.text(); // consume body
        return this.errorSnapshot(config.id);
      }
      const body = (await res.json()) as CostReportResponse;
      let used = 0;
      for (const bucket of body.data ?? []) {
        for (const r of bucket.results ?? []) {
          const amt = parseFloat(r?.amount ?? "0");
          used += amt;
        }
      }
      const limit = config.manualLimit ?? DEFAULT_LIMIT;
      const ratio = limit > 0 ? used / limit : 0;
      const status = ratio >= 1 ? "error" : ratio >= 0.8 ? "warning" : "ok";
      return {
        providerId: config.id,
        used,
        limit,
        unit: "USD",
        timestamp: Date.now(),
        status,
        displayName: config.displayName ?? "Anthropic",
      };
    } catch {
      return this.errorSnapshot(config.id);
    }
  }

  private errorSnapshot(providerId: string): UsageSnapshot {
    return {
      providerId,
      used: 0,
      limit: 1,
      unit: "USD",
      timestamp: Date.now(),
      status: "error",
      displayName: "Anthropic",
    };
  }
}
